pkg/pillar/cmd/zedrouter/zedrouter.go

```diff
@@ -1239,59 +1256,92 @@ func findBridge(bridgeName string) (*netlink.Bridge, error) {
 }
 
 // XXX Need additional logic for IPv6 underlays.
-func getUlAddrs(ctx *zedrouterContext,
-	ifnum int, appNum int,
-	status *types.UnderlayNetworkStatus,
-	netInstStatus *types.NetworkInstanceStatus) (string, string, error) {
-
-	log.Functionf("getUlAddrs(%d/%d)\n", ifnum, appNum)
+func getUlAddrs(ctx *zedrouterContext, ifnum int,
+	netInstStatus *types.NetworkInstanceStatus,
+	ulStatus *types.UnderlayNetworkStatus, appID uuid.UUID) (string, error) {
+	var err error
+	var mac net.HardwareAddr
+	networkID := netInstStatus.UUID
 
-	bridgeIPAddr := ""
-	appIPAddr := ""
+	if netInstStatus.Subnet.IP == nil ||
+		netInstStatus.DhcpRange.Start == nil {
+		log.Functionf("getUlAddrs(%d/%s) no subnet %s\n",
+			ifnum, networkID.String(), appID.String())
+		return "", nil
+	}
+	if ulStatus.Mac == "" {
+		log.Functionf("getUlAddrs(%d/%s) no mac %s\n",
+			ifnum, networkID.String(), appID.String())
+		return "", nil
+	}
+	log.Functionf("getUlAddrs(%d) for %s/%s\n", ifnum,
+		networkID.String(), appID.String())
 
-	// Allocate bridgeIPAddr based on BridgeMac
-	log.Functionf("getUlAddrs(%d/%d for %s) bridgeMac %s\n",
-		ifnum, appNum, netInstStatus.UUID.String(),
-		status.BridgeMac.String())
-	var err error
-	var addr string
-	addr, err = lookupOrAllocateIPv4(ctx, netInstStatus,
-		status.BridgeMac)
+	// XXX or change type of VifInfo.Mac to avoid parsing?
+	mac, err = net.ParseMAC(ulStatus.Mac)
 	if err != nil {
-		log.Errorf("getUlAddrs: Bridge IP address allocation failed %s\n", err)
-		return bridgeIPAddr, appIPAddr, err
+		errStr := fmt.Sprintf("parse Mac fail: %v\n", err)
+		log.Errorf("getUlAddrs(%s, %s): fail: %s\n",
+			networkID.String(), appID.String(), err)
+		return "", errors.New(errStr)
+	}
+
+	ipAddr := ""
+	// for static IP Address
+	if ulStatus.AppIPAddr != nil {
+		ipAddr = ulStatus.AppIPAddr.String()
+		// the IP Address, should not be in dhcpRange
+		if netInstStatus.DhcpRange.Contains(ulStatus.AppIPAddr) {
+			errStr := fmt.Sprintf("static IP(%s) is in DhcpRange(%s, %s)",
+				ipAddr, netInstStatus.DhcpRange.Start.String(),
+				netInstStatus.DhcpRange.End.String())
+			log.Errorf("getUlAddrs(%s, %s): fail: %s",
+				networkID.String(), appID.String(), errStr)
+			return "", errors.New(errStr)
+		}
+		// IP Address must be inside the subnet range
+		if !netInstStatus.Subnet.Contains(ulStatus.AppIPAddr) {
+			errStr := fmt.Sprintf("out of subnet range(%s)", ipAddr)
```

Comment:
Shouldn't the error say static IP(%s) outside of subnet; "static" is missing and makes sense to have the error string be similar to the one on line 1295